File: mini-diffusion-python/mini_diffusion/unet.py
"""
U-Net architecture for diffusion models.

Implements encoder-decoder with skip connections and time embeddings.
"""

import numpy as np
from typing import Optional

from .tensor import Tensor
from .layers import Linear, Conv2d, GroupNorm


def silu(x: Tensor) -> Tensor:
    """SiLU/Swish activation function."""
    return x.silu()


def timestep_embedding(timestep: int, dim: int, batch: int) -> Tensor:
    """
    Create sinusoidal timestep embedding.
    
    Args:
        timestep: Current timestep
        dim: Embedding dimension
        batch: Batch size
        
    Returns:
        Embedding tensor [batch, dim]
    """
    half_dim = dim // 2
    log_max = np.log(10000.0)
    
    emb = np.zeros((batch, dim))
    for i in range(half_dim):
        freq = np.exp(-log_max * i / half_dim)
        angle = timestep * freq
        emb[:, i] = np.sin(angle)
        emb[:, i + half_dim] = np.cos(angle)
    
    return Tensor(emb)


class ResBlock:
    """Residual block with time embedding."""
    
    def __init__(
        self,
        in_channels: int,
        out_channels: int,
        time_embed_dim: int,
        rng: Optional[np.random.Generator] = None
    ):
        """
        Initialize ResBlock.
        
        Args:
            in_channels: Number of input channels
            out_channels: Number of output channels
            time_embed_dim: Dimension of time embedding
            rng: Random number generator
        """
        if rng is None:
            rng = np.random.default_rng()
        
        self.in_channels = in_channels
        self.out_channels = out_channels
        
        self.norm1 = GroupNorm(32, in_channels)
        self.conv1 = Conv2d(in_channels, out_channels, 3, 1, 1, rng)
        self.norm2 = GroupNorm(32, out_channels)
        self.conv2 = Conv2d(out_channels, out_channels, 3, 1, 1, rng)
        self.time_proj = Linear(time_embed_dim, out_channels, rng)
        
        # Shortcut for channel mismatch
        if in_channels != out_channels:
            self.shortcut = Conv2d(in_channels, out_channels, 1, 1, 0, rng)
        else:
            self.shortcut = None
    
    def forward(self, x: Tensor, time_embed: Tensor) -> Tensor:
        """
        Forward pass.
        
        Args:
            x: Input tensor [batch, in_channels, height, width]
            time_embed: Time embedding [batch, time_embed_dim]
            
        Returns:
            Output tensor [batch, out_channels, height, width]
        """
        residual = self.shortcut(x) if self.shortcut else x
        
        # First conv block
        h = self.norm1(x)
        h = silu(h)
        h = self.conv1(h)
        
        # Add time embedding (project and broadcast)
        time_proj = self.time_proj(time_embed)  # [batch, out_channels]
        time_proj = Tensor(time_proj.data.reshape(time_proj.shape[0], time_proj.shape[1], 1, 1))
        h = h + time_proj
        
        # Second conv block
        h = self.norm2(h)
        h = silu(h)
        h = self.conv2(h)
        
        return h + residual
    
    def __call__(self, x: Tensor, time_embed: Tensor) -> Tensor:
        return self.forward(x, time_embed)
    
    def parameter_count(self) -> int:
        """Get number of parameters."""
        count = self.norm1.parameter_count() + self.conv1.parameter_count()
        count += self.norm2.parameter_count() + self.conv2.parameter_count()
        count += self.time_proj.parameter_count()
        if self.shortcut:
            count += self.shortcut.parameter_count()
        return count


class Downsample:
    """Downsample block using strided convolution."""
    
    def __init__(
        self,
        channels: int,
        rng: Optional[np.random.Generator] = None
    ):
        """
        Initialize Downsample.
        
        Args:
            channels: Number of channels
            rng: Random number generator
        """
        self.conv = Conv2d(channels, channels, 3, 2, 1, rng)
    
    def forward(self, x: Tensor) -> Tensor:
        """Forward pass - halves spatial dimensions."""
        return self.conv(x)
    
    def __call__(self, x: Tensor) -> Tensor:
        return self.forward(x)
    
    def parameter_count(self) -> int:
        """Get number of parameters."""
        return self.conv.parameter_count()


class Upsample:
    """Upsample block using nearest-neighbor + convolution."""
    
    def __init__(
        self,
        channels: int,
        rng: Optional[np.random.Generator] = None
    ):
        """
        Initialize Upsample.
        
        Args:
            channels: Number of channels
            rng: Random number generator
        """
        self.conv = Conv2d(channels, channels, 3, 1, 1, rng)
    
    def forward(self, x: Tensor) -> Tensor:
        """Forward pass - doubles spatial dimensions."""
        batch, channels, height, width = x.shape
        
        # Nearest neighbor 2x upsampling
        upsampled = np.zeros((batch, channels, height * 2, width * 2))
        upsampled[:, :, 0::2, 0::2] = x.data
        upsampled[:, :, 0::2, 1::2] = x.data
        upsampled[:, :, 1::2, 0::2] = x.data
        upsampled[:, :, 1::2, 1::2] = x.data
        
        return self.conv(Tensor(upsampled))
    
    def __call__(self, x: Tensor) -> Tensor:
        return self.forward(x)
    
    def parameter_count(self) -> int:
        """Get number of parameters."""
        return self.conv.parameter_count()


class UNet:
    """U-Net architecture for noise prediction in diffusion models."""
    
    def __init__(
        self,
        in_channels: int = 3,
        out_channels: int = 3,
        model_channels: int = 64,
        rng: Optional[np.random.Generator] = None
    ):
        """
        Initialize U-Net.
        
        Args:
            in_channels: Number of input channels
            out_channels: Number of output channels
            model_channels: Base channel count
            rng: Random number generator
        """
        if rng is None:
            rng = np.random.default_rng()
        
        self.in_channels = in_channels
        self.out_channels = out_channels
        self.model_channels = model_channels
        self.time_embed_dim = model_channels * 4
        
        # Input projection
        self.input_conv = Conv2d(in_channels, model_channels, 3, 1, 1, rng)
        
        # Time embedding MLP
        self.time_embed1 = Linear(model_channels, self.time_embed_dim, rng)
        self.time_embed2 = Linear(self.time_embed_dim, self.time_embed_dim, rng)
        
        # Encoder
        self.encoder_res1 = ResBlock(model_channels, model_channels, self.time_embed_dim, rng)
        self.down1 = Downsample(model_channels, rng)
        self.encoder_res2 = ResBlock(model_channels, model_channels * 2, self.time_embed_dim, rng)
        self.down2 = Downsample(model_channels * 2, rng)
        
        # Middle
        self.mid_res1 = ResBlock(model_channels * 2, model_channels * 2, self.time_embed_dim, rng)
        self.mid_res2 = ResBlock(model_channels * 2, model_channels * 2, self.time_embed_dim, rng)
        
        # Decoder (with skip connections)
        self.up1 = Upsample(model_channels * 2, rng)
        self.decoder_res1 = ResBlock(model_channels * 4, model_channels, self.time_embed_dim, rng)
        self.up2 = Upsample(model_channels, rng)
        self.decoder_res2 = ResBlock(model_channels * 2, model_channels, self.time_embed_dim, rng)
        
        # Output
        self.out_norm = GroupNorm(32, model_channels)
        self.out_conv = Conv2d(model_channels, out_channels, 3, 1, 1, rng)
    
    def forward(self, x: Tensor, timestep: int) -> Tensor:
        """
        Forward pass.
        
        Args:
            x: Input tensor [batch, in_channels, height, width]
            timestep: Current timestep
            
        Returns:
            Predicted noise [batch, out_channels, height, width]
        """
        batch = x.dim(0)
        
        # Time embedding
        t = timestep_embedding(timestep, self.model_channels, batch)
        t = self.time_embed1(t)
        t = silu(t)
        t = self.time_embed2(t)
        
        # Input projection
        h = self.input_conv(x)
        
        # Encoder with skip connections
        h1 = self.encoder_res1(h, t)      # [batch, model_channels, H, W]
        h2 = self.down1(h1)                # [batch, model_channels, H/2, W/2]
        h3 = self.encoder_res2(h2, t)     # [batch, model_channels*2, H/2, W/2]
        h4 = self.down2(h3)                # [batch, model_channels*2, H/4, W/4]
        
        # Middle
        mid = self.mid_res1(h4, t)
        mid = self.mid_res2(mid, t)
        
        # Decoder with skip connections
        up = self.up1(mid)                 # [batch, model_channels*2, H/2, W/2]
        up = self._concat(up, h3)          # [batch, model_channels*4, H/2, W/2]
        up = self.decoder_res1(up, t)      # [batch, model_channels, H/2, W/2]
        
        up = self.up2(up)                  # [batch, model_channels, H, W]
        up = self._concat(up, h1)          # [batch, model_channels*2, H, W]
        up = self.decoder_res2(up, t)      # [batch, model_channels, H, W]
        
        # Output
        out = self.out_norm(up)
        out = silu(out)
        out = self.out_conv(out)
        
        return out
    
    def _concat(self, a: Tensor, b: Tensor) -> Tensor:
        """Concatenate tensors along channel dimension."""
        return Tensor(np.concatenate([a.data, b.data], axis=1))
    
    def __call__(self, x: Tensor, timestep: int) -> Tensor:
        return self.forward(x, timestep)
    
    def parameter_count(self) -> int:
        """Get total number of parameters."""
        count = self.input_conv.parameter_count()
        count += self.time_embed1.parameter_count() + self.time_embed2.parameter_count()
        count += self.encoder_res1.parameter_count() + self.down1.parameter_count()
        count += self.encoder_res2.parameter_count() + self.down2.parameter_count()
        count += self.mid_res1.parameter_count() + self.mid_res2.parameter_count()
        count += self.up1.parameter_count() + self.decoder_res1.parameter_count()
        count += self.up2.parameter_count() + self.decoder_res2.parameter_count()
        count += self.out_norm.parameter_count() + self.out_conv.parameter_count()
        return count

File: mini-diffusion-java/src/main/java/com/minidiffusion/UNet.java
package com.minidiffusion;

import java.util.Random;

/**
 * U-Net architecture for diffusion models.
 */
public class UNet {

    /**
     * Residual block with time embedding.
     */
    public static class ResBlock {
        private final Layers.GroupNorm norm1;
        private final Layers.Conv2d conv1;
        private final Layers.GroupNorm norm2;
        private final Layers.Conv2d conv2;
        private final Layers.Linear timeProj;
        private final Layers.Conv2d shortcut;
        private final int inChannels;
        private final int outChannels;

        public ResBlock(int inChannels, int outChannels, int timeEmbedDim, Random rng) {
            this.inChannels = inChannels;
            this.outChannels = outChannels;

            this.norm1 = new Layers.GroupNorm(32, inChannels);
            this.conv1 = new Layers.Conv2d(inChannels, outChannels, 3, 1, 1, rng);
            this.norm2 = new Layers.GroupNorm(32, outChannels);
            this.conv2 = new Layers.Conv2d(outChannels, outChannels, 3, 1, 1, rng);
            this.timeProj = new Layers.Linear(timeEmbedDim, outChannels, rng);

            // Shortcut for channel mismatch
            if (inChannels != outChannels) {
                this.shortcut = new Layers.Conv2d(inChannels, outChannels, 1, 1, 0, rng);
            } else {
                this.shortcut = null;
            }
        }

        public Tensor forward(Tensor x, Tensor timeEmbed) {
            Tensor residual = shortcut != null ? shortcut.forward(x) : x;

            // First conv block
            Tensor h = norm1.forward(x);
            h = silu(h);
            h = conv1.forward(h);

            // Add time embedding (project and broadcast)
            Tensor timeProj2d = timeProj.forward(timeEmbed);
            h = addTimeEmbed(h, timeProj2d);

            // Second conv block
            h = norm2.forward(h);
            h = silu(h);
            h = conv2.forward(h);

            // Add residual
            return h.add(residual);
        }

        private Tensor silu(Tensor x) {
            Tensor result = x.clone();
            for (int i = 0; i < x.size(); i++) {
                double val = x.data[i];
                result.data[i] = val / (1.0 + Math.exp(-val));
            }
            return result;
        }

        private Tensor addTimeEmbed(Tensor x, Tensor timeEmbed) {
            // x: [batch, channels, height, width]
            // timeEmbed: [batch, channels]
            int batch = x.dim(0);
            int channels = x.dim(1);
            int height = x.dim(2);
            int width = x.dim(3);

            Tensor result = x.clone();
            for (int b = 0; b < batch; b++) {
                for (int c = 0; c < channels; c++) {
                    double te = timeEmbed.get(b, c);
                    for (int h = 0; h < height; h++) {
                        for (int w = 0; w < width; w++) {
                            result.set(result.get(b, c, h, w) + te, b, c, h, w);
                        }
                    }
                }
            }
            return result;
        }

        public int parameterCount() {
            int count = norm1.parameterCount() + conv1.parameterCount();
            count += norm2.parameterCount() + conv2.parameterCount();
            count += timeProj.parameterCount();
            if (shortcut != null) {
                count += shortcut.parameterCount();
            }
            return count;
        }
    }

    /**
     * Downsample block using strided convolution.
     */
    public static class Downsample {
        private final Layers.Conv2d conv;

        public Downsample(int channels, Random rng) {
            this.conv = new Layers.Conv2d(channels, channels, 3, 2, 1, rng);
        }

        public Tensor forward(Tensor x) {
            return conv.forward(x);
        }

        public int parameterCount() {
            return conv.parameterCount();
        }
    }

    /**
     * Upsample block using nearest-neighbor upsampling + convolution.
     */
    public static class Upsample {
        private final Layers.Conv2d conv;

        public Upsample(int channels, Random rng) {
            this.conv = new Layers.Conv2d(channels, channels, 3, 1, 1, rng);
        }

        public Tensor forward(Tensor x) {
            // Nearest neighbor 2x upsampling
            int batch = x.dim(0);
            int channels = x.dim(1);
            int height = x.dim(2);
            int width = x.dim(3);

            Tensor upsampled = Tensor.zeros(batch, channels, height * 2, width * 2);
            for (int b = 0; b < batch; b++) {
                for (int c = 0; c < channels; c++) {
                    for (int h = 0; h < height; h++) {
                        for (int w = 0; w < width; w++) {
                            double val = x.get(b, c, h, w);
                            upsampled.set(val, b, c, h * 2, w * 2);
                            upsampled.set(val, b, c, h * 2, w * 2 + 1);
                            upsampled.set(val, b, c, h * 2 + 1, w * 2);
                            upsampled.set(val, b, c, h * 2 + 1, w * 2 + 1);
                        }
                    }
                }
            }

            return conv.forward(upsampled);
        }

        public int parameterCount() {
            return conv.parameterCount();
        }
    }

    // U-Net components
    private final int inChannels;
    private final int outChannels;
    private final int modelChannels;
    private final int timeEmbedDim;

    private final Layers.Conv2d inputConv;
    private final Layers.Linear timeEmbed1;
    private final Layers.Linear timeEmbed2;

    // Encoder
    private final ResBlock encoderRes1;
    private final Downsample down1;
    private final ResBlock encoderRes2;
    private final Downsample down2;

    // Middle
    private final ResBlock midRes1;
    private final ResBlock midRes2;

    // Decoder
    private final Upsample up1;
    private final ResBlock decoderRes1;
    private final Upsample up2;
    private final ResBlock decoderRes2;

    private final Layers.GroupNorm outNorm;
    private final Layers.Conv2d outConv;

    public UNet(int inChannels, int outChannels, int modelChannels, Random rng) {
        this.inChannels = inChannels;
        this.outChannels = outChannels;
        this.modelChannels = modelChannels;
        this.timeEmbedDim = modelChannels * 4;

        // Input projection
        this.inputConv = new Layers.Conv2d(inChannels, modelChannels, 3, 1, 1, rng);

        // Time embedding MLP
        this.timeEmbed1 = new Layers.Linear(modelChannels, timeEmbedDim, rng);
        this.timeEmbed2 = new Layers.Linear(timeEmbedDim, timeEmbedDim, rng);

        // Encoder
        this.encoderRes1 = new ResBlock(modelChannels, modelChannels, timeEmbedDim, rng);
        this.down1 = new Downsample(modelChannels, rng);
        this.encoderRes2 = new ResBlock(modelChannels, modelChannels * 2, timeEmbedDim, rng);
        this.down2 = new Downsample(modelChannels * 2, rng);

        // Middle
        this.midRes1 = new ResBlock(modelChannels * 2, modelChannels * 2, timeEmbedDim, rng);
        this.midRes2 = new ResBlock(modelChannels * 2, modelChannels * 2, timeEmbedDim, rng);

        // Decoder (with skip connections, so double channels)
        this.up1 = new Upsample(modelChannels * 2, rng);
        this.decoderRes1 = new ResBlock(modelChannels * 4, modelChannels, timeEmbedDim, rng);
        this.up2 = new Upsample(modelChannels, rng);
        this.decoderRes2 = new ResBlock(modelChannels * 2, modelChannels, timeEmbedDim, rng);

        // Output
        this.outNorm = new Layers.GroupNorm(32, modelChannels);
        this.outConv = new Layers.Conv2d(modelChannels, outChannels, 3, 1, 1, rng);
    }

    /**
     * Forward pass through U-Net.
     */
    public Tensor forward(Tensor x, int timestep) {
        int batch = x.dim(0);

        // Time embedding
        Tensor t = timestepEmbedding(timestep, modelChannels, batch);
        t = timeEmbed1.forward(t);
        t = silu(t);
        t = timeEmbed2.forward(t);

        // Input projection
        Tensor h = inputConv.forward(x);

        // Encoder with skip connections
        Tensor h1 = encoderRes1.forward(h, t); // [batch, modelChannels, H, W]
        Tensor h2 = down1.forward(h1); // [batch, modelChannels, H/2, W/2]
        Tensor h3 = encoderRes2.forward(h2, t); // [batch, modelChannels*2, H/2, W/2]
        Tensor h4 = down2.forward(h3); // [batch, modelChannels*2, H/4, W/4]

        // Middle
        Tensor mid = midRes1.forward(h4, t);
        mid = midRes2.forward(mid, t);

        // Decoder with skip connections
        Tensor up = up1.forward(mid); // [batch, modelChannels*2, H/2, W/2]
        up = concat(up, h3); // [batch, modelChannels*4, H/2, W/2]
        up = decoderRes1.forward(up, t); // [batch, modelChannels, H/2, W/2]

        up = up2.forward(up); // [batch, modelChannels, H, W]
        up = concat(up, h1); // [batch, modelChannels*2, H, W]
        up = decoderRes2.forward(up, t); // [batch, modelChannels, H, W]

        // Output
        Tensor out = outNorm.forward(up);
        out = silu(out);
        out = outConv.forward(out);

        return out;
    }

    /**
     * Sinusoidal timestep embedding.
     */
    private Tensor timestepEmbedding(int timestep, int dim, int batch) {
        int halfDim = dim / 2;
        double logMax = Math.log(10000.0);

        Tensor emb = Tensor.zeros(batch, dim);
        for (int b = 0; b < batch; b++) {
            for (int i = 0; i < halfDim; i++) {
                double freq = Math.exp(-logMax * i / halfDim);
                double angle = timestep * freq;
                emb.set(Math.sin(angle), b, i);
                emb.set(Math.cos(angle), b, i + halfDim);
            }
        }
        return emb;
    }

    private Tensor silu(Tensor x) {
        Tensor result = x.clone();
        for (int i = 0; i < x.size(); i++) {
            double val = x.data[i];
            result.data[i] = val / (1.0 + Math.exp(-val));
        }
        return result;
    }

    /**
     * Concatenate two tensors along the channel dimension.
     */
    private Tensor concat(Tensor a, Tensor b) {
        int batch = a.dim(0);
        int ca = a.dim(1);
        int cb = b.dim(1);
        int height = a.dim(2);
        int width = a.dim(3);

        Tensor result = Tensor.zeros(batch, ca + cb, height, width);

        // Copy a
        for (int n = 0; n < batch; n++) {
            for (int c = 0; c < ca; c++) {
                for (int h = 0; h < height; h++) {
                    for (int w = 0; w < width; w++) {
                        result.set(a.get(n, c, h, w), n, c, h, w);
                    }
                }
            }
        }

        // Copy b
        for (int n = 0; n < batch; n++) {
            for (int c = 0; c < cb; c++) {
                for (int h = 0; h < height; h++) {
                    for (int w = 0; w < width; w++) {
                        result.set(b.get(n, c, h, w), n, ca + c, h, w);
                    }
                }
            }
        }

        return result;
    }

    public int parameterCount() {
        int count = inputConv.parameterCount();
        count += timeEmbed1.parameterCount() + timeEmbed2.parameterCount();
        count += encoderRes1.parameterCount() + down1.parameterCount();
        count += encoderRes2.parameterCount() + down2.parameterCount();
        count += midRes1.parameterCount() + midRes2.parameterCount();
        count += up1.parameterCount() + decoderRes1.parameterCount();
        count += up2.parameterCount() + decoderRes2.parameterCount();
        count += outNorm.parameterCount() + outConv.parameterCount();
        return count;
    }
}
